### DISCUSSION TITLE
Вывод "шахматки" для оперативного сравнения цен в HTML

### DISCUSSION THREAD

**Discussion by Vadimgrig2012** [totumonline/totum-ru-issues-and-discussions#354] [Как это сделать?] [answered]
Добрый день всем! Ситуация следующая: есть простая таблица с полями наименований товара, наименованиями поставщиков и предлагаемыми поставщиками ценами на товар. По условию поставщиков с предложением цены на 1 позицию товара может быть несколько. Таблица в виде:
![27a8d828-0659-49bf-8e35-e02084ae619d](https://user-images.githubusercontent.com/85877793/233591512-faa4c13c-98c2-4c84-81ea-3e695f96d9e9.jpg)

Необходимо получить для оперативного анализа цен предлагаемых поставщиками на каждую заявленную позицию нечто:
![Без имени](https://user-images.githubusercontent.com/85877793/233591627-c343e715-5151-472f-95a3-48bf8237d47e.png)

Средствами разворота таблицы данный функционал реализовать не получается. На выходе совсем не то. 
Пытаюсь данный функционал реализовать через linkToDataHtml, но никак не могу понять саму логику в коде (какие функции использовать и в какой последовательности). 
Я понимаю, что мне мне нужно вытащить значения цен из асс.массива и вставить их в колонки, где номер ячейки соответствует ключам "postavshik" и "naimenovanie", но вот в голове никак не укладывается, что использовать и с чего начать.

Есть кусок кода, который формирует HTML на рисунке ниже:

```
= : linkToDataHtml(title: "*"; html: $row; width: 1000; height: 700)
           row: textByTemplate(template: "test"; data: $data)
                   data: rowCreate(field: "title" = $val; field: "body" = $val2)
                            val: rowCreate(field: "template" = "test_title"; field: "data" = $header)
                            val2: rowCreate(field: "template" = "test_body"; field: "data" = $body)
                                    header: selectRowList(table: $#ntn; field: 'postavshchik')
                                    body: selectRowList(table: $#ntn; field: 'naimenovanie'; field: 'cena')
```

![8a076e99-e8e5-4ed3-9d08-d1cc314d655d](https://user-images.githubusercontent.com/85877793/233595861-5af58104-ab8f-4cb1-8c3a-526caf48e598.jpg)

Помогите пожалуйста разобраться. Может кто сталкивался с такого рода задачей.

**Comment by erpic-ru** [comment]
Нужно ли видеть все предложения по товарам от всех поставщиков на одном экране? 
Как вариант - вывести в таблицу список товаров, максимальное и минимальное предложение по цене, а также количество предложений. В панель ценовых предложений (или отдельной колонкой) вывести название поставщика, а по клику на количество сделать переход в модалку с простой таблицей со списком всех предложений с фильтром по товару.

**Reply by erpic-ru** [reply]
Есть ещё вариант захардкодить поставщиков в колонках :) Либо дождаться выхода динамических колонок. Кажется, оно уже в работе

**Reply by Vadimgrig2012** [reply]
> Нужно ли видеть все предложения по товарам от всех поставщиков на одном экране? Как вариант - вывести в таблицу список товаров, максимальное и минимальное предложение по цене, а также количество предложений. В панель ценовых предложений (или отдельной колонкой) вывести название поставщика, а по клику на количество сделать переход в модалку с простой таблицей со списком всех предложений с фильтром по товару.

Примерно такой же вариант, который вы предлагаете, уже реализован, но там немного по другому: таблица товаров с количеством и хар-ками (запрос от заказчика) ->последнее поле в данной таблице типа "кнопка" (на ней динамически отображается количество предложений при совпадении артикулов в данной таблице и таблице, в которую вносятся предложения от поставщиков)->при нажатии на кнопку отрабатывают префильтры и открывается окно с таблицей, в котором отображаются все предложения к конкретно данной позиции. Так же есть общая таблица, куда вносятся все предложения от поставщиков(там и соотношения по артикулам заказчика отрабатываются и еще куча всего, но это скорее для глубокой аналитики) Как-то так в общем. Но руководство просит отчет без лишней воды, которая не нужна для оперативного мониторинга и конкретно в вышеописанном формате: "Что б на кнопку жамкнул и оно все выскочило")) По этому ответ на вопрос"Нужно ли видеть все предложения по товарам от всех поставщиков на одном экране?" в конкретно моей ситуации однозначно Да.

**Reply by Vadimgrig2012** [reply]
Выхода динамических колонок очень жду, но пока этого нет - пытаюсь достичь рез-тат доступными методами.

**Comment by poluninatm** [comment, marked as answer]
<img width="582" alt="Снимок экрана 2023-04-21 в 15 26 20" src="https://user-images.githubusercontent.com/2241864/233635057-59e0b7e2-b8ec-4431-bfec-7e9ebebe8d3a.png">


Расчет кода поля для формирования таблицы в панельке:

```
= : rowCreate(field: "settings" = json`{"columns":$columns, "headRow": $vendorNames}`; field: "data"=$data)
	vendorNames: listAdd(list: json`["Наименование"]`; list: $uniqVendor)
	columns: listNumberRange(min: 0; max: $uniqVendorCount; step: 1)
		uniqVendorCount:  listCount(list: $uniqVendor) + 1
		
data: listReplace(list: $uniqProduct; action: $vals; value: "product")
	vals: listAdd(list: json`[$#product]`; list: $vendors)
		vendors: listReplace(list: $uniqVendor; action: $prices[0][cena]; value: "vendor")
			prices: listFilter(list: $sel; key: "postavshchik" =$#vendor; key: "tovar"=$#product)

~uniqVendor: listUniq(list: $sel[[postavshchik]])
~uniqProduct: listUniq(list: $sel[[tovar]])
~sel: selectRowList(table: $#ntn; sfield: 'tovar'; sfield: 'postavshchik'; field: 'cena'; order: 'id' asc)
```

По клику можно вызвать html-таблицу:

<img width="801" alt="Снимок экрана 2023-04-21 в 15 31 10" src="https://user-images.githubusercontent.com/2241864/233636596-67ee3795-7b37-43eb-8985-67674dc3d174.png">

```
= : linkToDataHtml(title: "*"; html: $table)
table: str`$style++"<table id='price'><thead><tr>"+$trhead+"</tr></thead><tbody>"+$tbody+"</tbody></table>"`
style: "<style>	#price td, #price th{		border: 1px solid black;		padding: 5px;		}</style>"

trhead: listJoin(list: $trHeads)
	trHeads: listReplace(list: $#nfv[settings][headRow]; action: str`"<th>"+$#val+"</th>"`; value: "val")
	tbody: listJoin(list: $rows)
		rows: listReplace(list: $#nfv[data]; action: $row; value: "tds")
			row: str`"<tr>"+$tr+"</tr>"`
				tr: listJoin(list: $tds; str: "")
					tds: listReplace(list: $#tds; action: $td; value: "td")
						td: str`"<td>"+$#td+"</td>"`
```

Если делать кнопку без панельки - $#nfv нужно заменить на расчет кода для панельки (с тильдой!)

**Reply by Vadimgrig2012** [reply]
Спасибо вам огромное. Буду разбирать.
